File: web/src/api/discovery.ts
import { api } from "./client";
import type { components } from "./schema";

export type LastFmSettings = components["schemas"]["LastFmSettingsResponse"];
export type DiscoverySuggestion =
  components["schemas"]["DiscoverySuggestionResponse"];
export type DiscoveryStats = components["schemas"]["DiscoveryStatsResponse"];

type ScanResult =
  | { success: true; newSuggestions: number; totalPending: number }
  | { success: false; error: string };

// --- Settings ---

export async function getSettings(): Promise<LastFmSettings | null> {
  const { data, error } = await api.GET("/discovery/settings");
  if (error) return null;
  return data;
}

export async function connectLastfm(
  username: string,
  apiKey: string,
): Promise<LastFmSettings | null> {
  const { data, error } = await api.POST("/discovery/connect", {
    body: { username, api_key: apiKey },
  });
  if (error) return null;
  return data;
}

export async function disconnectLastfm(): Promise<boolean> {
  const { error } = await api.POST("/discovery/disconnect");
  return !error;
}

export async function updateSettings(
  updates: Partial<LastFmSettings & { api_key?: string }>,
): Promise<LastFmSettings | null> {
  const { data, error } = await api.PUT("/discovery/settings", {
    body: updates as Record<string, unknown>,
  });
  if (error) return null;
  return data;
}

// --- Scan ---

export async function runScan(): Promise<ScanResult> {
  const { data, error } = await api.POST("/discovery/scan");
  if (error) {
    return { success: false, error: "Failed to run discovery scan" };
  }
  return {
    success: true,
    newSuggestions: data.new_suggestions,
    totalPending: data.total_pending,
  };
}

export async function generateSimilarTracksPlaylist(): Promise<ScanResult> {
  const { data, error } = await api.POST("/discovery/playlists/similar-tracks");
  if (error) {
    return {
      success: false,
      error: "Failed to generate similar tracks playlist",
    };
  }
  return {
    success: true,
    newSuggestions: data.new_suggestions,
    totalPending: data.total_pending,
  };
}

// --- Suggestions ---

export async function listSuggestions(
  status?: components["schemas"]["SuggestionStatus"] | null,
): Promise<DiscoverySuggestion[]> {
  const { data, error } = await api.GET("/discovery/suggestions", {
    params: { query: { status: status ?? null } },
  });
  if (error) return [];
  return data.items;
}

export async function approveSuggestion(
  id: string,
): Promise<DiscoverySuggestion | null> {
  const { data, error } = await api.POST(
    "/discovery/suggestions/{suggestion_id}/approve",
    { params: { path: { suggestion_id: id } } },
  );
  if (error) return null;
  return data;
}

export async function rejectSuggestion(id: string): Promise<boolean> {
  const { error } = await api.POST(
    "/discovery/suggestions/{suggestion_id}/reject",
    { params: { path: { suggestion_id: id } } },
  );
  return !error;
}

export async function bulkApprove(suggestionIds?: string[]): Promise<number> {
  const { data, error } = await api.POST(
    "/discovery/suggestions/bulk-approve",
    {
      body: { suggestion_ids: suggestionIds ?? null },
    },
  );
  if (error) return 0;
  return (data as Record<string, number>)?.approved ?? 0;
}

// --- Discover Playlist ---

type DiscoverPlaylistResult =
  | {
      success: true;
      approved: number;
      playlistTracks: number;
      playlistPath: string | null | undefined;
    }
  | { success: false; error: string };

export async function createDiscoverPlaylist(): Promise<DiscoverPlaylistResult> {
  const { data, error } = await api.POST("/discovery/playlists/discover");
  if (error) {
    return { success: false, error: "Failed to create Discover playlist" };
  }
  return {
    success: true,
    approved: data.approved,
    playlistTracks: data.playlist_tracks,
    playlistPath: data.playlist_path,
  };
}

// --- Clear ---

export async function clearSuggestions(): Promise<number> {
  const { data, error } = await api.POST("/discovery/suggestions/clear");
  if (error) return 0;
  return (data as Record<string, number>)?.cleared ?? 0;
}

// --- Stats ---

export async function getStats(): Promise<DiscoveryStats | null> {
  const { data, error } = await api.GET("/discovery/stats");
  if (error) return null;
  return data;
}

File: packages/api/src/yubal_api/api/routes/discovery.py
"""Discovery endpoints for Last.fm-powered music discovery."""

from uuid import UUID

from fastapi import APIRouter, status

from yubal_api.api.deps import DiscoveryServiceDep
from yubal_api.db.discovery import SuggestionStatus
from yubal_api.schemas.discovery import (
    BulkApproveRequest,
    DiscoverPlaylistResponse,
    DiscoveryScanResponse,
    DiscoveryStatsResponse,
    DiscoverySuggestionListResponse,
    DiscoverySuggestionResponse,
    LastFmConnectRequest,
    LastFmSettingsResponse,
    LastFmSettingsUpdate,
)

router = APIRouter(prefix="/discovery", tags=["discovery"])


@router.get("/settings", response_model=LastFmSettingsResponse | None)
def get_discovery_settings(
    service: DiscoveryServiceDep,
) -> LastFmSettingsResponse | None:
    """Get Last.fm discovery settings."""
    settings = service.get_settings()
    if settings is None:
        return None
    return LastFmSettingsResponse.model_validate(settings)


@router.put("/settings", response_model=LastFmSettingsResponse)
def update_discovery_settings(
    data: LastFmSettingsUpdate,
    service: DiscoveryServiceDep,
) -> LastFmSettingsResponse:
    """Update Last.fm discovery settings."""
    settings = service.update_settings(data.model_dump(exclude_unset=True))
    return LastFmSettingsResponse.model_validate(settings)


@router.post("/connect", response_model=LastFmSettingsResponse)
def connect_lastfm(
    data: LastFmConnectRequest,
    service: DiscoveryServiceDep,
) -> LastFmSettingsResponse:
    """Connect a Last.fm account."""
    settings = service.connect(data.username, data.api_key)
    return LastFmSettingsResponse.model_validate(settings)


@router.post("/disconnect", status_code=status.HTTP_204_NO_CONTENT)
def disconnect_lastfm(
    service: DiscoveryServiceDep,
) -> None:
    """Disconnect Last.fm and clear suggestions."""
    service.disconnect()


@router.post("/scan", response_model=DiscoveryScanResponse)
async def run_discovery_scan(
    service: DiscoveryServiceDep,
) -> DiscoveryScanResponse:
    """Run a discovery scan."""
    new_count = service.run_scan()
    stats = service.get_stats()
    return DiscoveryScanResponse(
        new_suggestions=new_count,
        total_pending=stats["pending"],
    )


@router.get("/suggestions", response_model=DiscoverySuggestionListResponse)
def list_suggestions(
    service: DiscoveryServiceDep,
    status: SuggestionStatus | None = None,
) -> DiscoverySuggestionListResponse:
    """List discovery suggestions."""
    items = service.get_suggestions(status=status)
    return DiscoverySuggestionListResponse(
        items=[DiscoverySuggestionResponse.model_validate(s) for s in items]
    )


@router.get("/stats", response_model=DiscoveryStatsResponse)
def get_discovery_stats(
    service: DiscoveryServiceDep,
) -> DiscoveryStatsResponse:
    """Get discovery statistics."""
    return DiscoveryStatsResponse(**service.get_stats())


RESPONSE_MODEL = DiscoverySuggestionResponse


@router.post("/suggestions/{suggestion_id}/approve", response_model=RESPONSE_MODEL)
def approve_suggestion(
    suggestion_id: UUID,
    service: DiscoveryServiceDep,
) -> DiscoverySuggestionResponse:
    """Approve a suggestion and enqueue download."""
    suggestion = service.approve_suggestion(suggestion_id)
    if suggestion is None:
        from fastapi import HTTPException

        raise HTTPException(status_code=404, detail="Suggestion not found")
    return DiscoverySuggestionResponse.model_validate(suggestion)


@router.post("/suggestions/{suggestion_id}/reject", response_model=RESPONSE_MODEL)
def reject_suggestion(
    suggestion_id: UUID,
    service: DiscoveryServiceDep,
) -> DiscoverySuggestionResponse:
    """Reject a suggestion."""
    suggestion = service.reject_suggestion(suggestion_id)
    if suggestion is None:
        from fastapi import HTTPException

        raise HTTPException(status_code=404, detail="Suggestion not found")
    return DiscoverySuggestionResponse.model_validate(suggestion)


@router.post("/suggestions/bulk-approve", response_model=dict)
def bulk_approve_suggestions(
    data: BulkApproveRequest,
    service: DiscoveryServiceDep,
) -> dict:
    """Bulk approve suggestions."""
    count = service.bulk_approve(data.suggestion_ids)
    return {"approved": count}


@router.post("/suggestions/clear", response_model=dict)
def clear_suggestions(
    service: DiscoveryServiceDep,
) -> dict:
    """Delete all discovery suggestions."""
    count = service.clear_suggestions()
    return {"cleared": count}


@router.post("/playlists/similar-tracks", response_model=DiscoveryScanResponse)
def generate_similar_tracks_playlist(
    service: DiscoveryServiceDep,
    top_tracks_limit: int = 10,
    similar_tracks_per_track: int = 3,
) -> DiscoveryScanResponse:
    """Generate a similar-tracks playlist (Discover Weekly-style)."""
    new_count = service.generate_similar_tracks_playlist(
        top_tracks_limit=top_tracks_limit,
        similar_tracks_per_track=similar_tracks_per_track,
    )
    stats = service.get_stats()
    return DiscoveryScanResponse(
        new_suggestions=new_count,
        total_pending=stats["pending"],
    )


@router.post("/playlists/discover", response_model=DiscoverPlaylistResponse)
async def create_discover_playlist(
    service: DiscoveryServiceDep,
) -> DiscoverPlaylistResponse:
    """Approve all pending and create a Discover playlist M3U."""
    result = await service.create_discover_playlist()
    return DiscoverPlaylistResponse(**result)
